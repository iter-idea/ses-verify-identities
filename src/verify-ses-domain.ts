import { CfnOutput, Fn, RemovalPolicy, Stack } from 'aws-cdk-lib';
import { CnameRecord, HostedZone, IHostedZone, MxRecord, TxtRecord } from 'aws-cdk-lib/aws-route53';
import { ITopic, Topic } from 'aws-cdk-lib/aws-sns';
import { AwsCustomResource, PhysicalResourceId } from 'aws-cdk-lib/custom-resources';
import { EnvironmentPlaceholders } from 'aws-cdk-lib/cx-api';
import { Construct } from 'constructs';
import { generateSesPolicyForCustomResource } from './helper';

export type NotificationType = 'Bounce' | 'Complaint' | 'Delivery';

const REGIONS_WITH_LOCAL_DKIM = ['eu-south-1'];

/**
 * @struct
 */
export interface IVerifySesDomainProps {
  /**
   * A domain name to be used for the SES domain identity, e.g. 'sub-domain.example.org'
   */
  readonly domainName: string;
  /**
   * A hosted zone name to be used for retrieving the Route53 hosted zone for adding new record, e.g. 'example.org'.
   * If you also provide hostedZoneId, it is assumed that these values are correct and no lookup happens.
   * @default same as domainName
   */
  readonly hostedZoneName?: string;
  /**
   * Optional: A hosted zone id to be used for retrieving the Route53 hosted zone for adding new records. Providing an
   * id will skip the hosted zone lookup.
   */
  readonly hostedZoneId?: string;
  /**
   * Whether to automatically add a TXT record to the hosed zone of your domain. This only works if your domain is
   * managed by Route53. Otherwise disable it.
   * @default true
   */
  readonly addTxtRecord?: boolean;
  /**
   * Whether to automatically add a MX record to the hosted zone of your domain. This only works if your domain is
   * managed by Route53. Otherwise disable it.
   * @default true
   */
  readonly addMxRecord?: boolean;
  /**
   * Whether to automatically add DKIM records to the hosted zone of your domain. This only works if your domain is
   * managed by Route53. Otherwise disable it.
   * @default true
   */
  readonly addDkimRecords?: boolean;
  /**
   * An SNS topic where bounces, complaints or delivery notifications can be sent to. If none is provided, a new topic
   * will be created and used for all different notification types.
   * @default new topic will be created
   */
  readonly notificationTopic?: ITopic;
  /**
   * Select for which notification types you want to configure a topic.
   * @default [Bounce, Complaint]
   */
  readonly notificationTypes?: NotificationType[];
  /**
   * Whether to DESTROY or RETAIN the domain on removal.
   *
   * @default DESTROY
   */
  readonly removalPolicy?: RemovalPolicy;
}

/**
 * A construct to verify a SES domain identity. It initiates a domain verification and can automatically create
 * appropriate records in Route53 to verify the domain. Also, it's possible to attach a notification topic for bounces,
 * complaints or delivery notifications.
 *
 * @example
 *
 * new VerifySesDomain(this, 'SesDomainVerification', {
 *   domainName: 'example.org'
 * });
 */
export class VerifySesDomain extends Construct {
  /**
   * The SNS topic where bounces, complaints or delivery notifications can be sent to.
   */
  public readonly notificationTopic: ITopic;

  constructor(parent: Construct, name: string, props: IVerifySesDomainProps) {
    super(parent, name);

    const {
      domainName,
      hostedZoneName,
      hostedZoneId,
      notificationTopic,
      notificationTypes,
      addTxtRecord,
      addMxRecord,
      addDkimRecords,
      removalPolicy
    } = props;

    const verifyDomainIdentity = this.verifyDomainIdentity(domainName, removalPolicy);
    this.notificationTopic = this.createTopicOrUseExisting(domainName, verifyDomainIdentity, notificationTopic);
    this.addTopicToDomainIdentity(domainName, this.notificationTopic, notificationTypes);

    const zone = this.getHostedZone({ name: hostedZoneName || domainName, id: hostedZoneId });
    if (!zone) throw new Error('Can not determine hosted zone. Provide a hostedZoneName or hostedZoneId.');

    if (isTrueOrUndefined(addTxtRecord)) {
      const txtRecord = this.createTxtRecordLinkingToSes(zone, domainName, verifyDomainIdentity);
      txtRecord.node.addDependency(verifyDomainIdentity);
    }

    if (isTrueOrUndefined(addMxRecord)) {
      const mxRecord = this.createMxRecord(zone, domainName);
      mxRecord.node.addDependency(verifyDomainIdentity);
    }

    if (isTrueOrUndefined(addDkimRecords)) {
      const verifyDomainDkim = this.initDkimVerification(domainName);
      verifyDomainDkim.node.addDependency(verifyDomainIdentity);
      const region = Stack.of(this).region;
      this.addDkimRecords(verifyDomainDkim, zone, domainName, region);
    }
  }

  private verifyDomainIdentity(domainName: string, removalPolicy?: RemovalPolicy): AwsCustomResource {
    return new AwsCustomResource(this, 'VerifyDomainIdentity', {
      onCreate: {
        service: 'SES',
        action: 'verifyDomainIdentity',
        parameters: {
          Domain: domainName
        },
        physicalResourceId: PhysicalResourceId.fromResponse('VerificationToken')
      },
      onUpdate: {
        service: 'SES',
        action: 'verifyDomainIdentity',
        parameters: {
          Domain: domainName
        },
        physicalResourceId: PhysicalResourceId.fromResponse('VerificationToken')
      },
      onDelete:
        removalPolicy === RemovalPolicy.RETAIN
          ? undefined
          : {
              service: 'SES',
              action: 'deleteIdentity',
              parameters: {
                Identity: domainName
              }
            },
      policy: generateSesPolicyForCustomResource('VerifyDomainIdentity', 'DeleteIdentity')
    });
  }

  private getHostedZone({ name, id }: { name?: string; id?: string }): IHostedZone | null {
    // if name and id are available, we can save a lookup call
    if (name && id) return HostedZone.fromHostedZoneAttributes(this, 'Zone', { hostedZoneId: id, zoneName: name });
    if (name) return HostedZone.fromLookup(this, 'Zone', { domainName: name });

    return null;
  }

  private createTxtRecordLinkingToSes(zone: IHostedZone, domainName: string, verifyDomainIdentity: AwsCustomResource) {
    return new TxtRecord(this, 'SesVerificationRecord', {
      zone,
      recordName: `_amazonses.${domainName}`,
      values: [verifyDomainIdentity.getResponseField('VerificationToken')]
    });
  }

  private createMxRecord(zone: IHostedZone, domainName: string) {
    return new MxRecord(this, 'SesMxRecord', {
      zone,
      recordName: domainName,
      values: [
        {
          hostName: Fn.sub(`inbound-smtp.${EnvironmentPlaceholders.CURRENT_REGION}.amazonaws.com`),
          priority: 10
        }
      ]
    });
  }

  private initDkimVerification(domainName: string) {
    return new AwsCustomResource(this, 'VerifyDomainDkim', {
      onCreate: {
        service: 'SES',
        action: 'verifyDomainDkim',
        parameters: {
          Domain: domainName
        },
        physicalResourceId: PhysicalResourceId.of(domainName + '-verify-domain-dkim')
      },
      onUpdate: {
        service: 'SES',
        action: 'verifyDomainDkim',
        parameters: {
          Domain: domainName
        },
        physicalResourceId: PhysicalResourceId.of(domainName + '-verify-domain-dkim')
      },
      policy: generateSesPolicyForCustomResource('VerifyDomainDkim')
    });
  }

  private addDkimRecords(verifyDomainDkim: AwsCustomResource, zone: IHostedZone, domainName: string, region: string) {
    [0, 1, 2].forEach(val => {
      const dkimToken = verifyDomainDkim.getResponseField(`DkimTokens.${val}`);
      const cnameRecord = new CnameRecord(this, 'SesDkimVerificationRecord' + val, {
        zone,
        recordName: `${dkimToken}._domainkey.${domainName}`,
        domainName: REGIONS_WITH_LOCAL_DKIM.includes(region)
          ? `${dkimToken}.dkim.${region}.amazonses.com`
          : `${dkimToken}.dkim.amazonses.com`
      });
      cnameRecord.node.addDependency(verifyDomainDkim);
    });
  }

  private createTopicOrUseExisting(
    domainName: string,
    verifyDomainIdentity: AwsCustomResource,
    existingTopic?: ITopic
  ): ITopic {
    const topic = existingTopic ?? new Topic(this, 'SesNotificationTopic');
    new CfnOutput(this, domainName + 'SesNotificationTopic', {
      value: topic.topicArn,
      description: 'SES notification topic for ' + domainName
    });
    topic.node.addDependency(verifyDomainIdentity);
    return topic;
  }

  private addTopicToDomainIdentity(domainName: string, topic: ITopic, notificationTypes?: NotificationType[]) {
    if (notificationTypes?.length) {
      notificationTypes.forEach(type => {
        this.addSesNotificationTopicForIdentity(domainName, type, topic);
      });
    } else {
      // by default, notify on errors like a bounce or complaint
      this.addSesNotificationTopicForIdentity(domainName, 'Bounce', topic);
      this.addSesNotificationTopicForIdentity(domainName, 'Complaint', topic);
    }
  }

  private addSesNotificationTopicForIdentity(
    identity: string,
    notificationType: NotificationType,
    notificationTopic: ITopic
  ): void {
    const addTopic = new AwsCustomResource(this, `Add${notificationType}Topic-${identity}`, {
      onCreate: {
        service: 'SES',
        action: 'setIdentityNotificationTopic',
        parameters: {
          Identity: identity,
          NotificationType: notificationType,
          SnsTopic: notificationTopic.topicArn
        },
        physicalResourceId: PhysicalResourceId.of(`${identity}-set-${notificationType}-topic`)
      },
      policy: generateSesPolicyForCustomResource('SetIdentityNotificationTopic')
    });

    addTopic.node.addDependency(notificationTopic);
  }
}

function isTrueOrUndefined(prop?: boolean): boolean {
  return prop === undefined || prop;
}
