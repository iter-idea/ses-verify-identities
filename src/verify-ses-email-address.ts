import { Construct } from '@aws-cdk/core';
import { AwsCustomResource, PhysicalResourceId } from '@aws-cdk/custom-resources';
import { generateSesPolicyForCustomResource } from './helper';

/**
 * @struct
 */
export interface IVerifySesEmailAddressProps {
  /**
   * The email address to be verified, e.g. 'hello@example.org'.
   */
  readonly emailAddress: string;
  /**
   * An optional AWS region to validate the email address.
   * @default The custom resource will be created in the stack region
   */
  readonly region?: string;
}

/**
 * A construct to verify an SES email address identity. It initiates a verification so that SES sends a verification email to the desired email address. This means the owner of the email address still needs to act by clicking the link in the verification email.
 *
 * @example
 *
 * new VerifySesEmailAddress(this, 'SesEmailVerification', {
 *   emailAddress: 'hello@example.org'
 * });
 *
 */
export class VerifySesEmailAddress extends Construct {
  constructor(parent: Construct, name: string, props: IVerifySesEmailAddressProps) {
    super(parent, name);

    const emailAddress = props.emailAddress;
    const region = props.region;

    new AwsCustomResource(this, 'VerifyEmailIdentity' + emailAddress, {
      onCreate: {
        service: 'SES',
        action: 'verifyEmailIdentity',
        parameters: {
          EmailAddress: emailAddress,
        },
        physicalResourceId: PhysicalResourceId.of('verify-' + emailAddress),
        region,
      },
      onDelete: {
        service: 'SES',
        action: 'deleteIdentity',
        parameters: {
          Identity: emailAddress,
        },
        region,
      },
      policy: generateSesPolicyForCustomResource('VerifyEmailIdentity', 'DeleteIdentity'),
    });
  }
}
