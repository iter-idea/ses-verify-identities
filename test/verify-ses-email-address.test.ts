import * as cdk from 'aws-cdk-lib';
import { RemovalPolicy } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { VerifySesEmailAddress } from '../src';

describe('verify-ses-email-address', () => {

  test('ensure custom resource exists to verify email address', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');
    const emailAddress = 'hello@example.org';

    new VerifySesEmailAddress(stack, 'VerifyExampleEmail', {
      emailAddress: emailAddress,
    });

    // Then
    const template = Template.fromStack(stack);
    template.resourceCountIs('Custom::AWS', 1);

    // ensure properties are as expected
    template.hasResourceProperties('Custom::AWS', {
      Create: Match.serializedJson(Match.objectLike({
        service: 'SES',
        action: 'verifyEmailIdentity',
        parameters: {
          EmailAddress: emailAddress,
        },
      })),
      Delete: Match.serializedJson({
        service: 'SES',
        action: 'deleteIdentity',
        parameters: {
          Identity: emailAddress,
        },
      }),
    });
  });

  test('ensure custom resource exists in a custom region to verify email address', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');
    const emailAddress = 'hello@example.org';
    const region = 'us-east-1';

    new VerifySesEmailAddress(stack, 'VerifyExampleEmail', {
      emailAddress: emailAddress,
      region,
    });

    // Then
    const template = Template.fromStack(stack);
    template.resourceCountIs('Custom::AWS', 1);

    // ensure properties are as expected
    template.hasResourceProperties('Custom::AWS', {
      Create: Match.serializedJson(Match.objectLike({
        service: 'SES',
        action: 'verifyEmailIdentity',
        parameters: {
          EmailAddress: emailAddress,
        },
        region,
      })),
      Delete: Match.serializedJson({
        service: 'SES',
        action: 'deleteIdentity',
        parameters: {
          Identity: emailAddress,
        },
        region,
      }),
    });
  });


  test('ensure custom resource is retained if removal policy is set', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');
    const emailAddress = 'hello@example.org';

    new VerifySesEmailAddress(stack, 'VerifyExampleEmail', {
      emailAddress: emailAddress,
      removalPolicy: RemovalPolicy.RETAIN,
    });

    // Then
    const template = Template.fromStack(stack);
    template.resourceCountIs('Custom::AWS', 1);

    // ensure create properties are as expected
    template.hasResourceProperties('Custom::AWS', {
      Create: Match.serializedJson(Match.objectLike({
        service: 'SES',
        action: 'verifyEmailIdentity',
        parameters: {
          EmailAddress: emailAddress,
        },
      })),
    });
  });

  test('ensure matches snapshot', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');
    const emailAddress = 'hello@example.org';

    new VerifySesEmailAddress(stack, 'VerifyExampleEmail', {
      emailAddress: emailAddress,
    });

    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
  });

  test('ensure matches snapshot with removal policy', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');
    const emailAddress = 'hello@example.org';

    new VerifySesEmailAddress(stack, 'VerifyExampleEmail', {
      emailAddress: emailAddress,
      removalPolicy: RemovalPolicy.RETAIN,
    });

    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
  });
});
