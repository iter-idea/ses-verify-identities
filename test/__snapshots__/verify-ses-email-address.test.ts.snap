// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verify-ses-email-address ensure matches snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "68b4ffa1c39cb3733535725f85311791c09eab53b7ab8efa5152e68f8abdb005.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "VerifyExampleEmailVerifyEmailIdentityhelloexampleorg61ABC7A2": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "VerifyExampleEmailVerifyEmailIdentityhelloexampleorgCustomResourcePolicyE8AAB538",
      ],
      "Properties": Object {
        "Create": "{\\"service\\":\\"SES\\",\\"action\\":\\"verifyEmailIdentity\\",\\"parameters\\":{\\"EmailAddress\\":\\"hello@example.org\\"},\\"physicalResourceId\\":{\\"id\\":\\"verify-hello@example.org\\"}}",
        "Delete": "{\\"service\\":\\"SES\\",\\"action\\":\\"deleteIdentity\\",\\"parameters\\":{\\"Identity\\":\\"hello@example.org\\"}}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "VerifyExampleEmailVerifyEmailIdentityhelloexampleorgCustomResourcePolicyE8AAB538": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ses:VerifyEmailIdentity",
                "ses:DeleteIdentity",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "VerifyExampleEmailVerifyEmailIdentityhelloexampleorgCustomResourcePolicyE8AAB538",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`verify-ses-email-address ensure matches snapshot with removal policy 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "68b4ffa1c39cb3733535725f85311791c09eab53b7ab8efa5152e68f8abdb005.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "VerifyExampleEmailVerifyEmailIdentityhelloexampleorg61ABC7A2": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "VerifyExampleEmailVerifyEmailIdentityhelloexampleorgCustomResourcePolicyE8AAB538",
      ],
      "Properties": Object {
        "Create": "{\\"service\\":\\"SES\\",\\"action\\":\\"verifyEmailIdentity\\",\\"parameters\\":{\\"EmailAddress\\":\\"hello@example.org\\"},\\"physicalResourceId\\":{\\"id\\":\\"verify-hello@example.org\\"}}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "VerifyExampleEmailVerifyEmailIdentityhelloexampleorgCustomResourcePolicyE8AAB538": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ses:VerifyEmailIdentity",
                "ses:DeleteIdentity",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "VerifyExampleEmailVerifyEmailIdentityhelloexampleorgCustomResourcePolicyE8AAB538",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
